### PR TITLE
{cae}[intel/2019b] FreeFEM v4.5 (WIP)

### PR DESCRIPTION
**Pull request by boegel**
(created using `eb --new-pr`)
requires https://github.com/easybuilders/easybuild-easyblocks/pull/1969

still WIP because although the installation works, I'm hitting errors like this when trying a non-trivial example:
`Intel MKL FATAL ERROR: Cannot load symbol MKLMPI_Get_wrappers` (cfr. https://github.com/easybuilders/easybuild-framework/issues/1673)

### CODE REVIEW

**Comment by boegel** [comment]
I *think* the `Cannot load symbol MKLMPI_Get_wrappers` issue stems from the `PETSc` dependency, where it's linking to `libmkl_blacs*` for which there's only a static library (`.a`), not a dynamic library (`.so`), which is exactly the issue being discussed at https://software.intel.com/en-us/forums/intel-math-kernel-library/topic/590302 .

See also https://lists.mcs.anl.gov/pipermail/petsc-dev/2016-January/018651.html, it seems to be a known but unresolved issue in PETSc?

**Comment by prj-** [comment]
> I _think_ the `Cannot load symbol MKLMPI_Get_wrappers` issue stems from the `PETSc` dependency, where it's linking to `libmkl_blacs*` for which there's only a static library (`.a`), not a dynamic library (`.so`), which is exactly the issue being discussed at https://software.intel.com/en-us/forums/intel-math-kernel-library/topic/590302 .
> 
> See also https://lists.mcs.anl.gov/pipermail/petsc-dev/2016-January/018651.html, it seems to be a known but unresolved issue in PETSc?

I think this issue can be fixed by explicitly linking the MPI libraries, but I haven't face it for a while.
I'm sorry I have a hard time deciphering the EasyBuild logs, where is this error is showing exactly, please?

File: easybuild/easyconfigs/b/BemTool/BemTool-20180912.eb [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+name = 'BemTool'
+version = '20180912'
+local_commit = '900479a'
+
+homepage = 'https://github.com/xclaeys/BemTool'
+description = "BemTool is a C++ boundary element library."
+
+toolchain = SYSTEM
+
+source_urls = ['https://github.com/xclaeys/BemTool/archive/']
```
Comment (prj-) [inline easybuild/easyconfigs/b/BemTool/BemTool-20180912.eb]: Careful, this is not the BemTool version required by FreeFEM. It needs a special fork, cf. [3rdparty/getall](https://github.com/FreeFem/FreeFem-sources/blob/develop/3rdparty/getall#L201).

**Comment by akesandgren** [comment]
Eh, there are .so versions of all the libmkl_blacs_* so i'm not sure what you mean by this.

**Comment by boegel** [comment]
> > I _think_ the `Cannot load symbol MKLMPI_Get_wrappers` issue stems from the `PETSc` dependency, where it's linking to `libmkl_blacs*` for which there's only a static library (`.a`), not a dynamic library (`.so`), which is exactly the issue being discussed at https://software.intel.com/en-us/forums/intel-math-kernel-library/topic/590302 .
> > See also https://lists.mcs.anl.gov/pipermail/petsc-dev/2016-January/018651.html, it seems to be a known but unresolved issue in PETSc?
> 
> I think this issue can be fixed by explicitly linking the MPI libraries, but I haven't faced it for a while.
> I'm sorry I have a hard time deciphering the EasyBuild logs, where is this error showing exactly, please?

@prj- You won't find the `Cannot load symbol MKLMPI_Get_wrappers` errors in the Travis logs (since that only runs the test suite, it doesn't actually try to install or use the software that the easyconfig (`.eb`) file installs).

The error also doesn't pop up at all during the installation, only when running a non-trivial example with `mpirun FreeFem++-mpi test.edp`.

Do you have more information on what you mean by "explicitly linking the MPI libraries"?

**Comment by prj-** [comment]
Which example? I can deactivate ScaLAPACK within MUMPS just as Stefano did so that the error will go away, I think. Concerning the explicit linking, I remember that one fix I found for this problem was to just put the libmpi.a before the MKL BLACS/ScaLAPACK library. But this may be system specific.

**Comment by boegel** [comment]
@prj- I need to check whether I can share this specific input file (it's not mine), I'll get back to you on that...

I don't think forcibly linking to `libmpi.a` will help (and if it does, whether it's a good solution), since the `Cannot load symbol MKLMPI_Get_wrappers` seems to be caused by linking to a static library while the runtime expects a dynamic library to be found (see https://software.intel.com/en-us/forums/intel-math-kernel-library/topic/590302).

**Comment by prj-** [comment]
> @prj- I need to check whether I can share this specific input file (it's not mine), I'll get back to you on that...

If it's not possible:
1. just looking at the `sparams` or the command-line arguments should suffice
2. you can try to bypass the problem yourself with the additional argument `-mat_mumps_icntl_13 1`

> I don't think forcibly linking to `libmpi.a` will help (and if it does, whether it's a good solution), since the `Cannot load symbol MKLMPI_Get_wrappers` seems to be caused by linking to a static library while the runtime expects a dynamic library to be found (see https://software.intel.com/en-us/forums/intel-math-kernel-library/topic/590302).

I agree it's not a good solution, it's just the solution I came up with when I last encountered this issue.

**Comment by boegel** [comment]
> > @prj- I need to check whether I can share this specific input file (it's not mine), I'll get back to you on that...
> 
> If it's not possible:
> 
>     1. just looking at the `sparams` or the command-line arguments should suffice
> 
>     2. you can try to bypass the problem yourself with the additional argument `-mat_mumps_icntl_13 1`

Is that an argument to the `FreeFem++-mpi` command, or should I change the input file somehow?

There is no `sparams` in the input file, no mention of `arg` either.

**Comment by prj-** [comment]
You can just supply that on the command line. You can also add `-ksp_view` to make sure that in the log you have something like:
```
              ICNTL(13) (efficiency control):                         1
```
(I'll submit a PETSc MR, this is not the correct text inbetween the parenthesis)

**Comment by boegel** [comment]
> You can just supply that on the command line. You can also add `-ksp_view` to make sure that in the log you have something like:
> 
> ```
>               ICNTL(13) (efficiency control):                         1
> ```
> 
> (I'll submit a PETSc MR, this is not the correct text inbetween the parenthesis)

I tried this, but I'm hitting the exact same `Intel MKL FATAL ERROR: Cannot load symbol MKLMPI_Get_wrappers.` issue...

```
mpirun -np 12 FreeFem++-mpi cube.edp -mat_mumps_icntl_13 1 -ksp_view
```

Where can I find the log file you're referring to?

The input file I'm using is this one: https://gist.github.com/boegel/ae7b69b2651cade9c58f6c4a6290fe3e

**Comment by prj-** [comment]
Oh, I thought you were using MUMPS through PETSc, not MUMPS directly. You can forget about my options then. Maybe if you `load "PETSc"` just before MUMPS, it will link to the appropriate wrappers? Alternatively, I can give you the MUMPS + PETSc version of this script, so that you can bypass the error as originally proposed (not to mention that it will be more efficient).

**Review by jfgrimm — COMMENTED** [review]
As decided in issue #16330, we have deprecated the use of `True` to signify a system-toolchain dependency (#16384), in favour of the more intuitive `SYSTEM` template constant. Due to the change in the test suite, please run `eb --sync-pr-with-develop ${pr}` and update the PR to use `SYSTEM` instead.

**Comment by jfgrimm** [comment]
closing since this has gone stale

**Pull request closed without merging**